### PR TITLE
Webpack upgrade move out tests

### PR DESCRIPTION
**Pull request by EdwinGuzman**
Upgrading to webpack 4. Changes include:

* upgrades to other npm packages, also removing packages that are not needed
* updated syntax (some which may be annoying but we can configure later if we want to)
* moved tests outside of the `src` folder so that they won't be part of the build
* updated webpack to a recommended configuration for dev and prod

_Managed to reduce the main js file from 800kb to 140kb._

This branch works in the Registry Admin - https://github.com/NYPL-Simplified/registry_admin/commit/5792b7a7c71c3f4391061900ba972629781e970f

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -11,44 +11,43 @@
   "license": "Apache-2.0",
   "scripts": {
     "prepublish": "typings install && npm run prod",
-    "lint": "tslint -c tslint.json src/*.ts src/*.tsx src/**/*.ts src/**/*.tsx src/**/**/*.ts src/**/**/*.tsx && sass-lint -c .sass-lint.yml -v -q",
-    "test": "npm run lint && tsc && cp -r src/stylesheets lib && mocha --require lib/testHelper.js lib/__tests__/*.js lib/**/__tests__/*.js lib/**/**/__tests__/*.js",
+    "lint": "tslint -c tslint.json src/*.ts src/*.tsx src/components/*.ts src/components/*.tsx && sass-lint -c .sass-lint.yml -v -q",
+    "test": "npm run lint && tsc && cp -r src/stylesheets lib && mocha --require lib/__tests__/testHelper.js lib/__tests__/*.js",
     "test-file": "npm run lint && tsc && cp -r src/stylesheets lib && mocha --require lib/testHelper.js",
-    "dev": "rm -rf dist && NODE_ENV=development webpack --watch --progress --display-modules --config webpack.prod.config",
-    "prod": "rm -rf dist && NODE_ENV=production webpack -p --progress --display-modules --config webpack.prod.config"
```
Comment (EdwinGuzman) [inline package.json]: Webpack 4, by default, configures the `NODE_ENV` variable for you in their respective config files.

File: src/Demo.tsx [hunk 2]
```diff
@@ -14,7 +14,10 @@ export interface DemoState {
   infoText: string;
 }
 
-export default class Demo extends React.Component<void, DemoState> {
+export default class Demo extends React.Component<any, DemoState> {
+  state: DemoState;
+  setState: any;
```
Comment (EdwinGuzman) [inline src/Demo.tsx]: These are slightly annoying and I'm not sure how to best "fix" these yet. I think this is fine for now and could be looked at in a different PR.

File: src/stylesheets/colors.scss [hunk 3]
```diff
@@ -1,3 +1,5 @@
+$focuscolor: #0F465C;
```
Comment (EdwinGuzman) [inline src/stylesheets/colors.scss]: The entire `opds-web-client` package was deleted but this and the `basic-focus` mixin had to be moved into this repo.

File: typings.json [hunk 4]
```diff
@@ -5,18 +5,11 @@
...
     "node": "github:DefinitelyTyped/DefinitelyTyped/node/node.d.ts#86dbea8fc37d9473fee465da4f0a21bea4f8cbd9",
     "qs": "github:DefinitelyTyped/DefinitelyTyped/qs/qs.d.ts#7de6c3dd94feaeb21f20054b9f30d5dabc5efabd",
     "react": "registry:dt/react#0.14.0+20160630100702",
     "react-bootstrap": "github:NYPL-Simplified/DefinitelyTyped/react-bootstrap/react-bootstrap.d.ts#7d462d047559bc2587224e1ecdc3a043aa3e2626",
-    "react-dom": "registry:dt/react-dom#0.14.0+20160412154040",
-    "react-redux": "registry:dt/react-redux#4.4.0+20160703120921",
-    "react-router": "registry:dt/react-router#2.0.0+20160501155536",
-    "react-router/history": "registry:dt/react-router/history#2.0.0+20160316155526",
-    "redux": "registry:dt/redux#3.5.2+20160703092728",
-    "redux-thunk": "registry:dt/redux-thunk#2.1.0+20160703120921"
```
Comment (EdwinGuzman) [inline typings.json]: Unneeded typings are removed.

File: webpack.common.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  output: {
+    path: path.resolve(__dirname, "./dist"),
+    publicPath: "/",
+    filename: "reusable-components.js",
+    library: "ReusableComponents",
+    libraryTarget: "umd"
+  },
+
+  plugins: [
+    new CleanWebpackPlugin()
```
Comment (EdwinGuzman) [inline webpack.common.js]: Used to remove the `/dist` folder every time before building.

File: webpack.dev.js [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+      "./src/index.tsx"
+    ]
+  },
+
+  module: {
+    rules: [
+      {
+        test: /\.scss$/,
+        use: [
+          "style-loader",
```
Comment (EdwinGuzman) [inline webpack.dev.js]: Doing this removes the the need to import `reusable-components.css` in the demo. For the dev mode, it's fine to include the css _inside_ the js file. In production the recommended way is to have its own css file. We are not using the webpack dev server (which we should use) but doing this further makes using the dev server easier.

File: webpack.prod.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+const path = require("path");
+const MiniCssExtractPlugin = require("mini-css-extract-plugin");
+const merge = require('webpack-merge');
+
+const common = require('./webpack.common.js');
+
+module.exports = merge(common, {
+  mode: "production",
+
+  // devtool: "source-map",
```
Comment (EdwinGuzman) [inline webpack.prod.js]: I'm not sure if we want source maps in production for a components library so I commented it out, but it is a recommended practice. I removed it because it makes the build folder bigger.

File: webpack.prod.js [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+      filename: "reusable-components.css"
+    })
+  ],
+
+  module: {
+    rules: [
+      {
+        test: /\.scss$/,
+        use: [
+          MiniCssExtractPlugin.loader,
```
Comment (EdwinGuzman) [inline webpack.prod.js]: This creates the minified css file.

File: webpack.prod.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    ]
+  },
+
+  // When importing a module whose path matches one of the following, just
+  // assume a corresponding global variable exists and use that instead.
+  // This is important because it allows us to avoid bundling all of our
+  // dependencies, which allows browsers to cache those libraries between builds.
+  externals: {
+    'react-addons-test-utils': true,
+    'react': 'react',
```
Comment (EdwinGuzman) [inline webpack.prod.js]: For the production build, we do not want to include React, and since they are React components, it's ok to assume that they will be used in React projects.

**Comment by EdwinGuzman** [comment]
@gkallenberg  @aslagle you both don't have to review the PR if you don't want to, but I'm adding you both to see the changes needed for the upgrade to webpack 4. If you want to review, that's welcomed!

File: webpack.common.js [hunk 10]
```diff
@@ -0,0 +1,40 @@
+const path = require("path");
```
Comment (EdwinGuzman) [inline webpack.common.js]: Webpack recommends this pattern of having `common`, `dev`, and `prod` files and merging the needed configurations together.

File: __tests__/Fieldset-test.tsx [hunk 11]
```diff
@@ -1,7 +1,9 @@
+/// <reference path="../typings/globals/mocha/index.d.ts" />
```
Comment (aslagle) [inline __tests__/Fieldset-test.tsx]: What made this necessary?

Comment (aslagle) [reply __tests__/Fieldset-test.tsx]: Also, with moving the tests, I would probably match the directory structure under src, and we can ditch the underscores (I think maybe that was a React convention for indicating it wasn't part of the source?). So I'd go with `tests/components/Fieldset-test.tsx`.

File: index.html [hunk 12]
```diff
@@ -5,7 +5,6 @@
     <meta name="viewport" content="width=device-width, initial-scale=1.0">
     <title>Reusable Components</title>
     <link rel="shortcut icon" href="src/components/logo.png" type="image/x-icon" />
-    <link rel="stylesheet" type="text/css" href="dist/reusable-components.css">
```
Comment (aslagle) [inline index.html]: I haven't looked at this repo before so I'm not sure why this page exists, but why doesn't it need CSS anymore?

File: package.json [hunk 13]
```diff
@@ -11,44 +11,43 @@
   "license": "Apache-2.0",
   "scripts": {
     "prepublish": "typings install && npm run prod",
-    "lint": "tslint -c tslint.json src/*.ts src/*.tsx src/**/*.ts src/**/*.tsx src/**/**/*.ts src/**/**/*.tsx && sass-lint -c .sass-lint.yml -v -q",
-    "test": "npm run lint && tsc && cp -r src/stylesheets lib && mocha --require lib/testHelper.js lib/__tests__/*.js lib/**/__tests__/*.js lib/**/**/__tests__/*.js",
+    "lint": "tslint -c tslint.json src/*.ts src/*.tsx src/components/*.ts src/components/*.tsx && sass-lint -c .sass-lint.yml -v -q",
```
Comment (aslagle) [inline package.json]: I think the `src/**/*.ts` is more flexible in case we added another directory next to `components`. However, if this repo is all components maybe we should flatten it out and put everything directly under `src`.

File: package.json (hunk 1 above)
Comment (aslagle) [reply package.json]: What happened to `--watch`?

File: src/Demo.tsx [hunk 14]
```diff
@@ -14,7 +14,10 @@ export interface DemoState {
   infoText: string;
 }
 
-export default class Demo extends React.Component<void, DemoState> {
+export default class Demo extends React.Component<any, DemoState> {
```
Comment (aslagle) [inline src/Demo.tsx]: Again, not sure what this is for. But it sounds like something that should maybe be in a separate place and so it's not built into the published package?

File: src/Demo.tsx (hunk 2 above)
Comment (aslagle) [reply src/Demo.tsx]: Did you update the React typings too?

File: src/Demo.tsx [hunk 15]
```diff
@@ -76,15 +79,15 @@ export default class Demo extends React.Component<void, DemoState> {
...
+  makeColorList(colors: {}): Element {
     return (
       <ul className="clearfix" style={{margin: "0", backgroundColor: `${this.state.lights ? "#fff" : "#000"}`}}>
         { Object.entries(colors).map(color => this.makeColorSwatch(color)) }
       </ul>
     );
   }
 
-  makeColorSwatch(info: string[]): JSX.Element {
+  makeColorSwatch(info: string[] | {}): Element {
```
Comment (aslagle) [inline src/Demo.tsx]: This change to the `info` type seems weird?

File: src/components/Header.tsx [hunk 16]
```diff
@@ -8,8 +8,12 @@ export interface HeaderProps {
   logOut?: string;
 }
 
+declare const require: any;
```
Comment (aslagle) [inline src/components/Header.tsx]: I don't see where `require` is used.

File: index.html (hunk 12 above)
Comment (aslagle) [reply index.html]: never mind, found it.

File: webpack.prod.js [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+  entry: {
+    app: [
+      "./src/stylesheets/app.scss",
+      "./src/components.tsx"
+    ]
+  },
+
+  plugins: [
+    // jsdom is required by opds-web-client for server rendering, but causes
+    // errors in the browser even if it is never used, so we ignore it:
```
Comment (aslagle) [inline webpack.prod.js]: Is this still needed?

File: package.json (hunk 1 above)
Comment (EdwinGuzman) [reply package.json]: We're not running a dev server so it's not needed at the moment.

File: src/Demo.tsx (hunk 2 above)
Comment (EdwinGuzman) [reply src/Demo.tsx]: No I did not. That's a different issue I wanted to take care of later by introducing `@typings` and wanted to just isolate updating webpack.

File: __tests__/Fieldset-test.tsx (hunk 11 above)
Comment (EdwinGuzman) [reply __tests__/Fieldset-test.tsx]: Sounds good. It was necessary because of the upgrade in `ts-lint`. I'm not exactly sure what it was but the config and how it builds is different.

File: index.html (hunk 12 above)
Comment (EdwinGuzman) [reply index.html]: It's a demo page Adriana created.

File: package.json (hunk 13 above)
Comment (EdwinGuzman) [reply package.json]: Good point. I'll leave that for Adriana to decide since she's working on the components for this repo.

File: src/Demo.tsx (hunk 14 above)
Comment (EdwinGuzman) [reply src/Demo.tsx]: Right. This is only included in the dev webpack build but in the production build it is not included (the entry field in the respective dev and prod files are different).

File: src/components/Header.tsx (hunk 16 above)
Comment (EdwinGuzman) [reply src/components/Header.tsx]: Line 17.

Comment (aslagle) [reply src/components/Header.tsx]: Ah, right. I wonder if we still need that...

File: webpack.prod.js (hunk 17 above)
Comment (aslagle) [reply webpack.prod.js]: You could update the comment either way.

File: __tests__/Fieldset-test.tsx (hunk 11 above)
Comment (EdwinGuzman) [reply __tests__/Fieldset-test.tsx]: `tsc` does create a `__tests__` folder for you so that may be the convention. To be consistent, I'm leaving the main test folder as `__tests__` too, but I moved the tests to `__tests_/components/....tsx`.

**Comment by EdwinGuzman** [comment]
With @aslagle 's help, the size of the css build went down from 291kb + 106kb (SVG) to 6.65kb. Bootstrap was moved to `dev-dependencies` and only included in the dev webpack build. The demo works as before but it is just not included in the production build. We are assuming that the parent app will also include the bootstrap dependency.

@adriana-alter this makes me realize that the components are heavily influenced by bootstrap and if we want to move away from it for this repo, we'd need some normalize and base styling. But, that's a big topic for another time.

File: src/stylesheets/demo.scss [hunk 18]
```diff
@@ -0,0 +1,2 @@
+@import "../../node_modules/bootstrap/dist/css/bootstrap.min.css";
```
Comment (EdwinGuzman) [inline src/stylesheets/demo.scss]: @adriana-alter just want to point out that the components still depend on the base styling from bootstrap. For the demo, it's needed.

File: package.json (hunk 1 above)
Comment (adriana-alter) [reply package.json]: We are running a dev server; I've been running `npm link` and `npm run dev` on this when I'm working on importing the components from to the interfaces, so I can make changes to this as I go.  Is removing `--watch` going to make it so I can't do that...?

File: index.html (hunk 12 above)
Comment (adriana-alter) [reply index.html]: It's a demo/local development page; much of the point of it is so that I can just see the components in isolation in the browser.  Definitely still needs CSS.

File: package.json (hunk 13 above)
Comment (adriana-alter) [reply package.json]: I do anticipate the repo being all components, but I like having components and stylesheets in separate folders to make it easier to find things.

File: src/stylesheets/demo.scss (hunk 18 above)
Comment (adriana-alter) [reply src/stylesheets/demo.scss]: My goal for this iteration is just to get away from react-bootstrap, not regular bootstrap.  But in the next pass...

File: package.json (hunk 1 above)
Comment (EdwinGuzman) [reply package.json]: That's not the same thing as a dev server. The dev server would actually start a server and serve you files on a port in localhost. This was not set up to do this from the beginning. You were using `npm run dev` before to build a local development build and then you went to the index file directly to see the changes. That is not the same as a dev server.

`--watch` watches the files as you work on them and rebuilds it but since we not using a dev server I don't think there's much point to it. We can revisit this later and build in an actual dev server.

Comment (EdwinGuzman) [reply package.json]: You can still work on this as you were before and run `npm run dev` whenever you update the package, but it's not really what you want.

File: index.html (hunk 12 above)
Comment (EdwinGuzman) [reply index.html]: It does need CSS but webpack handles that for the dev build.

File: package.json (hunk 1 above)
Comment (EdwinGuzman) [reply package.json]: I temporarily added it back in so it can help with your dev workflow but that should be updated to be better.

Comment (aslagle) [reply package.json]: I think watch is still useful without a dev server. It won't automatically load changes, but you can update by reloading the page instead of having to manually run the command.

Comment (adriana-alter) [reply package.json]: Agreed

Comment (EdwinGuzman) [reply package.json]: 👍 I'll leave it in then.

**Comment by EdwinGuzman** [comment]
@adriana-alter let me know if anything else needs to be updated for this PR.

**Comment by adriana-alter** [comment]
No, I think it's all set!

**Comment by aslagle** [comment]
I still want to look into a couple Typescript things quick before you merge.

File: __tests__/Fieldset-test.tsx (hunk 11 above)
Comment (aslagle) [reply __tests__/Fieldset-test.tsx]: I don't like this reference - it's a global dependency, so it needs to be loaded for all the tests and just happens to be in this one because it's the first test file. That means it could easily break if you add or remove tests or rename files. I think it would be better to put it in the compiler options in the tsconfig file, but I'm having trouble getting it to work and I think it may be a bug with Typescript 2.4.1. If you're planning to upgrade Typescript farther than that in a future PR, we could revisit this then.

**Comment by EdwinGuzman** [comment]
Sounds good. I do want to upgrade Typescript soon since I think it'll help reduce one "fix" I did in the Storybook PR #6 .

File: src/Demo.tsx [hunk 19]
```diff
@@ -78,21 +78,24 @@ export default class Demo extends React.Component<void, DemoState> {
...
+        className="clearfix"
+        style={{margin: "0", backgroundColor: `${this.state.lights ? "#fff" : "#000"}`}}
+      >
         { Object.entries(colors).map(color => this.makeColorSwatch(color)) }
       </ul>
     );
   }
 
-  makeColorSwatch(info: string[]): JSX.Element {
+  makeColorSwatch(info: string[] | {}): JSX.Element {
```
Comment (aslagle) [inline src/Demo.tsx]: Not a big deal since this is the demo, but I think you can fix this by declaring some types for the color constant, like 
```
interface ColorCodeMap {
  [key: string]: string];
}
interface ColorNameMap {
  [key: string]: ColorCodeMap;
}
```

I don't know what it's doing now since as far as I can tell the input is always a non-empty object, not a string array.

**Comment by EdwinGuzman** [comment]
@aslagle  @adriana-alter can review this PR again and also pull it into the registry admin or circulation web to test?

One change I made is I updated the npm script `prepublish` to `prepublishOnly` since `prepublish` is now deprecated. So now the production build will only run before deploying to npm (and not when doing npm install). The only drawback I see is if you want to link the package, you'd have to run `npm link` and `npm run prod` (`npm run dev` won't work since it's the dev build). So it may not be a continuous development process since you'd to run `npm run prod` every time you make a new change. But, this is also assuming that you run `npm run dev` to test the components in isolation and only link to test the components when it's in the other app.

Let me know what you think. I can update the script to `prepare` instead of `prepublish` so when you link it also gives the production build, but you'd still have to run `npm run prod` every time.

File: webpack.prod.js (hunk 7 above)
Comment (aslagle) [reply webpack.prod.js]: I'd say leave this in.

Comment (EdwinGuzman) [reply webpack.prod.js]: My only worry is that it creates a map file which is larger than the compiled source.
```
                Asset      Size  Chunks             Chunk Names
    reusable-components.css  2.03 KiB       0  [emitted]  app
reusable-components.css.map  2.81 KiB       0  [emitted]  app
     reusable-components.js   143 KiB       0  [emitted]  app
 reusable-components.js.map   495 KiB       0  [emitted]  app
```

**Comment by aslagle** [comment]
I think `npm link` should still work with the dev build - that's what I normally do in other repos. I don't think it requires the prod build to run first, even though that's what we were doing before. Is there something else different about this that makes it not work?

File: webpack.prod.js (hunk 7 above)
Comment (aslagle) [reply webpack.prod.js]: I think that's probably ok - the applications using this could choose to ignore those files if they don't need them.

**Comment by EdwinGuzman** [comment]
It's the way that I set up the dev and prod webpack configs. The production build has the css and js files that other apps look for when running `npm link`, but the dev build only outputs the js file which contains the css inside. It's common practice but not too helpful for when when we link.

**Comment by aslagle** [comment]
Ah I see. That could end up being annoying but we could always revisit it later.

**Comment by EdwinGuzman** [comment]
Yes, it can be annoying. @adriana-alter the local development workflow won't be changed, so let me know what you think.

**Comment by adriana-alter** [comment]
Looks good to me!
